### DISCUSSION TITLE
Custom scripts input validation on REST API route

### DISCUSSION THREAD

**Discussion by monnierj** [netbox-community/netbox#18229] [Ideas]
First of all, thanks for all the work done on Netbox!

Custom script input is properly validated on the WebUI (with a Django form), but on the REST API side of the feature, there's just a JSONField that validates JSON well-formedness and... that's all.

This is a real issue when ObjectVar are used on the script, since we obtain a number or a string instead of the expected model instance in the data dictionary.

Netbox users seems mitigates the issue in various ways (as in https://github.com/netbox-community/netbox/discussions/8344, https://github.com/netbox-community/netbox/discussions/16255), which generally implies to use a non-declared script variable, or to check script variable types before using them.

In my opinion, there are two things to be done:

- create a function similar to BaseScript.as_form(), which would return a proper Django Rest Framework serializer to validate the script input, and make it match the validation rules set in StringVar/ObjectVar/etc. on the Script object.
- add a "api_lookup_field" property on ObjectVars, to allow developpers to change which model field will be used to find a given model instance. By default we should use the `id` field, but using the `slug` could be interesting in some cases.

I volunteer to implement this feature if considered relevant.

However, this is a breaking change: existing scripts could break once input validation is applied on the REST API.

Should we implement said validation? And should make API input validation optional to begin, and later make it mandatory?

**Comment by matejv** [comment]
I'm also interested in running custom scripts via API and also see the current functionality as lacking.

There might be also a third way: 

In `ScriptInputSerializer` you could validate data by calling `script.as_form()` and pass input data then return `form.cleaned_data` and it will be passed to script. There would be a bit more work to transform any form validation errors in proper Django Rest Framework errors, but that should not be difficult. There is a similar approach in [runscript CLI command](https://github.com/netbox-community/netbox/blob/015ef25ca03291ab111c0e86169013f79269a57d/netbox/extras/management/commands/runscript.py#L77).

It feels a bit wrong to use django forms in django rest framework, but I think this approach would be by far the simplest.

This is a quick & dirty hack to illustrate what I ment:

```diff
diff --git a/netbox/extras/api/serializers_/scripts.py b/netbox/extras/api/serializers_/scripts.py
index 897ccf966..6486e2f41 100644
--- a/netbox/extras/api/serializers_/scripts.py
+++ b/netbox/extras/api/serializers_/scripts.py
@@ -65,6 +65,14 @@ class ScriptInputSerializer(serializers.Serializer):
     schedule_at = serializers.DateTimeField(required=False, allow_null=True)
     interval = serializers.IntegerField(required=False, allow_null=True)
 
+    def validate_data(self, value):
+        script = self.context['script'].python_class
+        script = script()
+        form = script.as_form(value, None)
+        if not form.is_valid():
+            raise serializers.ValidationError(form.errors.get_json_data())
+        return form.cleaned_data
+
     def validate_schedule_at(self, value):
         if value and not self.context['script'].scheduling_enabled:
             raise serializers.ValidationError(_("Scheduling is not enabled for this script."))
```

**Reply by monnierj** [reply]
That's a good point. We already have working form validation; re-using it would force the API-triggered script runs to follow exactly the same rules than in the WebUI.

There's still the issue of giving the developer the choice of the model field to use for `ObjectVar`s and `MultiObjectVar`s, but it can be handled before running the form validation.

**Comment by sjurtf** [comment]
I also have interests in this.

The option in the first post to add a property `api_lookup_field` on `ObjectVars` seems like a good idea.
In the case that I'm working on now, it would make sense to be able to use `name` to lookup instead of `id`.

**Comment by ross-cello** [comment]
Have also just encountered this problem (tested against v4.3.2 and v4.5.1). Came across this naturally while searching issues.

Has there been any movement toward raising a proper Issue?

**Comment by llamafilm** [comment]
This was fixed in the CLI by passing `form.cleaned_data` instead of `form.data` (#18612). I think the same approach would work here.  Put it in `ScriptViewSet` and make the API return a 400 error if some field doesn't pass valiation.
